pkg/network/state.go

```diff
@@ -477,6 +478,20 @@ func (ns *networkState) storeClosedConnections(conns []ConnectionStats) {
 					stateTelemetry.statsCookieCollisions.Inc()
 					// pick the latest one
 					if c.LastUpdateEpoch > client.closedConnections[i].LastUpdateEpoch {
+						if !isEmpty(c) && i >= client.emptyConnsIndex {
+							if i == client.emptyConnsIndex {
+								client.closedConnections[i] = c
+								client.emptyConnsIndex++
+								continue
+							}
+							emptyConn := client.closedConnections[client.emptyConnsIndex]
+							client.closedConnections[client.emptyConnsIndex] = c
+							client.closedConnections[i] = emptyConn
+							client.closedConnectionsKeys[c.Cookie] = client.emptyConnsIndex
+							client.closedConnectionsKeys[emptyConn.Cookie] = i
+							client.emptyConnsIndex++
+							continue
+						}
 						client.closedConnections[i] = c
```

Comment:
Won't we hit this for `isEmpty(c) || i < client.emptyConnsIndex`? That doesn't look right.

Reply:
If `isEmpty(c)` it doesn't matter if it replaces it, it will replace it for another empty connection which is okay. If `i < client.emptyConnsIndex` the index of the connection will remain the same since it was not an empty connection, so it will remain in the index it was before and just add the metrics of the merged connection. Is there a flaw to this logic that I'm not seeing?

Reply:
if `isEmpty(c)` is `true`, `i >= client.emptyConnsIndex` should also be `true`, otherwise we're replacing a non-empty connection with an empty one, possibly in the middle of the collection of non-empty connections? I think the case you're missing is when both clauses are true.

Reply:
i.e., `isEmpty(c)` and `i < client.emptyConnsIndex` are both `true`.

Reply:
if `i < client.emptyConnsIndex` it means that the connections with which `c` merged is not empty and the result of the `merge` would therefore be not be empty so we can just re-insert c in the index. I don't think that both `isEmpty(c)` and  `i < client.emptyConnsIndex` can be true.

Reply:
This branch actually means that the connection was not merged, so we're actually replacing the connection here.

Reply:
ah, got it